OrbitCore/SamplingProfiler.cpp

```diff
@@ -61,22 +61,22 @@ std::multimap<int, CallstackID> SamplingProfiler::GetCallstacksFromAddress(
 }
 
 //-----------------------------------------------------------------------------
-void SamplingProfiler::AddCallStack(CallstackEvent& callstack_event) {
+void SamplingProfiler::AddCallStack(CallstackEvent callstack_event) {
```

Comment:
Why pass argument by value here? It seems we copy/move on the caller side, then again in this function.

Reply:
reference leads to one extra-copy in this case: https://g3doc.corp.google.com/devtools/library_club/g3doc/totw/117.md?cl=head

Reply:
@dimitry- , @dpallotti , this still doesn't make a whole lot of sense to me. If the object being passed by value has some POD data, then we end up copying this static data twice. Shouldn't AddCallstack take in a rvalue reference as argument? This way we save a copy if the caller std::moves the argument.  If not, then the caller has to create a temporary which makes the extra copy explicit.

Reply:
You are correct about POD. Indeed, when the argument is fully POD you get "Clang-Tidy: std::move of the variable '...' of the trivially-constructible type '...' has no effect; remove std::move()".
But passing by const-ref or using the pass-by-valye-and-move pattern doesn't really make a significant difference with just a few POD fields. But you don't want many POD fields anyway, otherwise when you pass an r-value you are not able to move them, and be it that you copy once (passing an r-value reference you can't move from or a const-ref) or twice (passing by value), it's still too much copying. Instead you want to be able to move that data, by storing it inside the class with smart pointers, vectors, strings, etc, all of which can be moved. 
So the fact that you don't have too much POD data to copy regardless of whether you are copying or moving is pretty much an assumption. At this point, using pass-by-value-and-move is about adhering to a pattern that with a lot of movable data becomes quite relevant. If we always use it, then we also use it when we should use it.

You mention rvalue references. That would be the ideal parameter to pass r-values. But what if you want to pass an l-value (and make a copy of it)? Then you need the two overloads, one for the r-value, one for the l-value. That's an extra method and extra code. But what if you have two parameters (both of which to get into the class, are not just read)? Then you need four overloads, for the different combinations. And so on, that's not sustainable.
So what you do instead is have the parameter(s) by value and move from them. If you pass an l-value, the copy happens when you pass, and then you have the move in the method&mdash;if you were passing by const-ref, you would not have the move but you would still have the copy in the method. But if you pass an r-value, you have two moves and no copies&mdash;if you were passing by const-ref, you would not have the first move but then you would have to copy in the method, as you obviously can't move from a const-ref.
So l-value: one copy + one move instead of one copy. But r-value: **two moves instead of one copy**.

Reply:
> You mention rvalue references. That would be the ideal parameter to pass r-values. But what if you want to pass an l-value (and make a copy of it)? Then you need the two overloads, one for the r-value, one for the l-value.

Well in that case you can create a temporary object at the call site without having to declare an extra method.

> **the fact that the move is cheap is an assumption**.
This is exactly my point, yes. You have to know about the type you're passing, in some cases there might be expensive copies of static data.

> So l-value: one copy + one move instead of one copy. But r-value: **two moves instead of one copy**.

But if you have a r-value as argument, you have only one move.

See https://repl.it/repls/UltimateOrangeredWebpages#main.cpp

Reply:
> Well in that case you can create a temporary object at the call site without having to declare an extra method.

That gets very annoying very quickly in general.

> You have to know about the type you're passing, in some cases there might be expensive copies of static data.

Is there expensive that to pass in this case though? This is a proto. I would guess moves are cheap.